staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root.go

```diff
@@ -33,12 +34,21 @@ import (
 // GroupManager is an interface that allows dynamic mutation of the existing webservice to handle
 // API groups being added or removed.
 type GroupManager interface {
+	GroupLister
+
 	AddGroup(apiGroup metav1.APIGroup)
 	RemoveGroup(groupName string)
 	ServeHTTP(resp http.ResponseWriter, req *http.Request)
 	WebService() *restful.WebService
 }
 
+// GroupLister knows how to list APIGroups for discovery.
+type GroupLister interface {
+	// Groups returns APIGroups for discovery, filling in ServerAddressByClientCIDRs
+	// based on data in req.
+	Groups(ctx context.Context, req *http.Request) ([]metav1.APIGroup, error)
+}
```

Comment:
is there a reason for this interface? Why not just add the func directly to `GroupManager`? Does not look like a super useful abstraction.

Reply:
No reason at all. Just a pattern I saw around. (mostly in client-go). I was thinking under the lines:
"Ok, most people will use main interface methods, and only people deep in the stack will use extended one, so it kinda splits existing method to this extension type of thingy, where if you want you could have 2 implementations." 

Might be a lot of overthinking :) change?

Reply:
I remember now why its a separate interface :) in the API-extension server, we need just this single method. Easier to implement.